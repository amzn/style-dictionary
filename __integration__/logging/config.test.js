/*
 * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

const fs = require('fs-extra');
const StyleDictionary = require('../../index');
const {buildPath, cleanConsoleOutput} = require('../_constants');

// Spy on console.log and add all messages to an array
let consoleOutput = [];
const log = jest.spyOn(console, "log")
  .mockImplementation(message => consoleOutput.push(message))

/**
 * These integration tests will verify the behavior and logging at the *config*
 * level. These messages happen when `.extend()` is called to verify
 * proper configuration such as source being an array. This will also check
 * for collisions in source files and any errors that happen when parsing
 * and merging properties. This is the first of 3 phases of logging, the
 * next two are: platform and file.
 */
describe(`integration >`, () => {
  // before each test clear the mocked console.log and the output array
  beforeEach(() => {
    log.mockClear();
    consoleOutput = [];
  });

  describe(`logging >`, () => {
    describe(`config >`, () => {
      describe(`property value collisions`, () => {
        it(`should not throw, but notify users by default`, () => {
          StyleDictionary.extend({
            source: [
              // including a specific file twice will throw value collision warnings
              `__integration__/tokens/size/padding.json`,
              `__integration__/tokens/size/padding.json`
            ],
            platforms: {
            }
          });
          expect(consoleOutput.map(cleanConsoleOutput).join(`\n`)).toMatchSnapshot();
        });

        it(`should not show warnings if given higher log level`, () => {
          StyleDictionary.extend({
            logLevel: `error`,
            source: [
              // including a specific file twice will throw value collision warnings
              `__integration__/tokens/size/padding.json`,
              `__integration__/tokens/size/padding.json`
            ],
            platforms: {
            }
          });
          expect(consoleOutput.map(cleanConsoleOutput).join(`\n`)).toMatchSnapshot();
        });
      });
    });
  });
});

afterAll(() => {
  fs.emptyDirSync(buildPath);
});