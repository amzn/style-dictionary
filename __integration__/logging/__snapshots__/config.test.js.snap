// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration > logging > config > property value collisions should not show warnings if given higher log level 1`] = `
"
Property Value Collisions:
Collision detected at: size.padding.small! Original value: 0.5, New value: 0.5
Collision detected at: size.padding.small! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.small! Original value: true, New value: true
Collision detected at: size.padding.medium! Original value: 1, New value: 1
Collision detected at: size.padding.medium! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.medium! Original value: true, New value: true
Collision detected at: size.padding.large! Original value: 1, New value: 1
Collision detected at: size.padding.large! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.large! Original value: true, New value: true
Collision detected at: size.padding.xl! Original value: 1, New value: 1
Collision detected at: size.padding.xl! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.xl! Original value: true, New value: true

"
`;

exports[`integration > logging > config > property value collisions should not throw, but notify users by default 1`] = `
"
Property Value Collisions:
Collision detected at: size.padding.small! Original value: 0.5, New value: 0.5
Collision detected at: size.padding.small! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.small! Original value: true, New value: true
Collision detected at: size.padding.medium! Original value: 1, New value: 1
Collision detected at: size.padding.medium! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.medium! Original value: true, New value: true
Collision detected at: size.padding.large! Original value: 1, New value: 1
Collision detected at: size.padding.large! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.large! Original value: true, New value: true
Collision detected at: size.padding.xl! Original value: 1, New value: 1
Collision detected at: size.padding.xl! Original value: __integration__/tokens/size/padding.json, New value: __integration__/tokens/size/padding.json
Collision detected at: size.padding.xl! Original value: true, New value: true

"
`;
